### ISSUE TITLE
[Help please] How to debug "Invalid OAuth callback." error?

### ISSUE THREAD

**Issue by paul4156** [Shopify/shopify-node-api#308]
I am stuck on Shopify app installation with "Invalid OAuth callback." error in auth/callback stage.
How can I debug whether hmac or state causes this error?

The relevant code is as follows. Shop name should be fine (bc-prod-paul.myshopify.com).

```
    if (!validQuery(query, currentSession)) {
      throw new ShopifyErrors.InvalidOAuthError('Invalid OAuth callback.');
    }

...

function validQuery(query: AuthQuery, session: Session): boolean {
  return (
    validateHmac(query) &&
    validateShop(query.shop) &&
    safeCompare(query.state, session.state as string)
  );
}
```

Thanks!

**Comment by zakariamofaddel** [comment]
In my case most of the times this error popped out was because of this the last comparison made in the **validQuery** helper function `safeCompare(query.state, session.state as string)`.

The comparison was failing because the current session (which you get by using the Shopify util `loadCurrentSession`) had a different state than the one that was expected by Shopify. This happened because for some reason on installation the authentication route was being called twice in a span of less than 2 seconds, as a consequence the storeCallback was being called twice, overriding the original session that Shopify now uses to compare state.
So basically the `validQuery` function was comparing the first session's state to the session I had saved in my DB (which is the wrong one, result of the second call to the auth route).

I know this is very confusing, I suggest you console.log your way through the whole OAuth flow, to get a full picture of what is happening. Log every time your `storeCallback` and `loadCallback` fire and log the payload, so that you can understand which part fails the `validQuery` util.

**Comment by paul4156** [comment]
Thank you for this information and you are right. I have 2 server instances behind a load balancer. After removing one instance, the issue is resolved. Maybe the installation is called on both instances? That can perfectly explain the symptom. I will try to see if I can introduce a "lock" mechanism to prevent both instances firing the installation.

**Comment by chao-ding-awx** [comment]
@paul4156  Hi, I have exactly the same error here now.
Do you have a solution how to manage multi instance now?

**Comment by paul4156** [comment]
> @paul4156 Hi, I have exactly the same error here now. Do you have a solution how to manage multi instance now?

Yes. I have 2 web servers and was using in memory session storage. Changing to redis session storage resolved the problem. See https://github.com/Shopify/shopify-node-api/blob/main/docs/usage/customsessions.md

**Comment by hashhesh** [comment]
Hi @zakariamofaddel & @paul4156  , many thanks guys for useful explanation above and I believe I have the same issue as yours so if you don't mind can you provide me with insights of how have you solved yours.
Here is my findings to help you digest the issue:

1. I have `console.log` at the begging of `storeCallback` and at the begging of `loadCallback` and every time this 
method called the `state` is the same within `req.query`  so all fine.
and one more initiative thing state is same in both `storeCallback` and `loadCallback`, maybe there is 
override 😅 who knows. Till `validQuery`  function and exactly in `validateHmac` function the generated local
hmac is never ever the same hmac coming from `req.query` I have doubted in every step in validateHmac.
Even I have written my own validateHmac(only for test purpose) but unfortunately I still have no
idea when/how/why received hmac is never the same of local generated hmac. here is the functions i have used
for validating hmac and their output within `/auth/callback` route.

```javascript
function safeCompare(strA, strB) {
  if (typeof strA === typeof strB) {
    var buffA = void 0;
    var buffB = void 0;
    if (typeof strA === 'object' && typeof strB === 'object') {
      buffA = Buffer.from(JSON.stringify(strA));
      buffB = Buffer.from(JSON.stringify(strB));
    }
    else {
      buffA = Buffer.from(strA);
      buffB = Buffer.from(strB);
    }
    if (buffA.length === buffB.length) {
      return crypto.timingSafeEqual(buffA, buffB);
    }
  }
  else {
    throw new ShopifyErrors.SafeCompareError("Mismatched data types provided: " + typeof strA + " and " + typeof strB);
  }
  return false;
}

function validateHmac(query) {

  var hmac = query.hmac;
  var localHmac = generateLocalHmac(query);
  return safeCompare(hmac, localHmac);
}

function generateLocalHmac(_a) {
  var code = _a.code, timestamp = _a.timestamp, state = _a.state, shop = _a.shop, host = _a.host;
  var queryString = stringifyQuery(({
    code: code,
    timestamp: timestamp,
    state: state,
    shop: shop }, (host && { host: host })));
  return crypto.createHmac('sha256',  process.env.SHOPIFY_API_SECRET)
      .update(queryString)
      .digest('hex');
}

function stringifyQuery(query) {
  var orderedObj = Object.keys(query).map(key =>`${key}=${Array(query[key]).join(',')}`).join('&');
  return queryString.stringify(orderedObj);
}
```
What's make it very weired is if I try to console.log in `/auth/callback` route both `safeCompare` and `validateHmac`
```javascript
console.log('#before validateAuthCallback ** validateHMAC #', validateHmac(req.query));
// output -> false
```

```javascript
console.log('#before validateAuthCallback ** safeCompare result#', safeCompare(req.query.state, (await Shopify.Utils.loadCurrentSession(req, res, true)).state));
// output -> true
```
Please note: I have used shopify cli to generate this app. I have deployed my app to heroku and using
Heroku Data for Redis addon. Node v16.13.0, @shopify/shopify-api v3.0.0, 

That said, I have implemented `CustomSessionStorage` with redis and it's works fine. I don't know if i have covered all 
required information to help you guys know what's going on within  my code. Please if you would like to know
more information please let me know.

I highly appreciate your help.
Many thanks in advance

**Comment by devstein** [comment]
@hashhesh Did you ever figure this out? I'm running into the same issue

> the generated local hmac is never ever the same hmac coming from req.query

**Comment by paulz-swyftx** [comment]
@hashhesh @devstein Did you setup API_KEY and API_SECRET correctly? In your cases, hmac seems not correct. It is not related to the issue of running multiple server instances. Please run one server and verify the auth working first.

**Comment by SzuJanLin** [comment]
@devstein  Had the same problem and figure out it was my API_SECRET, thanks for finding this error
